File: src/components/DayNightController.cpp
#include "DayNightController.hpp"
#include "graphics/Light.hpp"
#include "core/Entity.hpp"
#include "core/Scene.hpp"
#include "input/Input.hpp"
#include "util/component_registry.hpp"

REGISTER_COMPONENT_CLASS(DayNightController, "dayNightController");

json_interpreter<DayNightController> DayNightController::s_properties{
	{ "daySkyColor",			&DayNightController::m_daySkyColor },
	{ "nightSkyColor",			&DayNightController::m_nightSkyColor },
	{ "dayLightColor",			&DayNightController::m_dayLightColor },
	{ "nightLightColor",		&DayNightController::m_nightLightColor },
	{ "dayLightIntensity",		&DayNightController::m_dayLightIntensity },
	{ "nightLightIntensity",	&DayNightController::m_nightLightIntensity }
};

DayNightController::DayNightController(Entity* parent) : Component(parent), m_dayTime(true), m_ambient(true), m_light(nullptr),
m_dayLightIntensity(0.0f), m_nightLightIntensity(0.0f) { }

void DayNightController::start_impl()
{
	m_light = getEntity()->getComponent<Light>();
	applyValues();
}

void DayNightController::update_impl()
{
	auto input = Input::instance();

	if (input->getKeyPressed(key_n)) {
		setDayTime(!getDayTime());
	}

	if (input->getKeyPressed(key_b)) {
		setAmbient(!getAmbient());
	}
}

void DayNightController::apply_json_impl(const nlohmann::json& json)
{
	Component::apply_json_impl(json);
	s_properties.interpret_all(this, json);
}

void DayNightController::setDayTime(bool val)
{
	if (val != m_dayTime) {
		m_dayTime = val;
		applyValues();
	}
}

void DayNightController::setAmbient(bool val)
{
	if (val != m_ambient) {
		m_ambient = val;
		applyValues();
	}
}

void DayNightController::applyValues()
{
	Scene* scene = getEntity()->getParentScene();

	if (scene) {
		glm::vec4 skyColor = m_ambient ? (m_dayTime ? m_daySkyColor : m_nightSkyColor) : glm::vec4(0.f);
		scene->setBackColor(skyColor);
		scene->setAmbientLight(skyColor);
	}
	
	if (m_light) {
		m_light->setColor(m_dayTime ? m_dayLightColor : m_nightLightColor);
		m_light->setIntensity(m_dayTime ? m_dayLightIntensity : m_nightLightIntensity);
	}
}

File: content/scripts/DayNightController.lua
local color = require("xmath.color")

DayNightController = Behaviour:new{
    dayTime = true,
    ambient = true,
    daySkyColor = color.black(),
    nightSkyColor = color.black(),
    dayLightColor = color.black(),
    nightLightColor = color.black(),
    dayLightIntensity = 0,
    nightLightIntensity = 0
}

function DayNightController:init()
    self.light = self:entity():getComponent("Light")
    self:applyValues()
end

function DayNightController:update()
    if Input.getKeyPressed("n") then
        self:setDayTime(not self.dayTime)
    end

    if Input.getKeyPressed("b") then
        self:setAmbient(not self.ambient)
    end
end

function DayNightController:setDayTime(val)
    if val ~= self.dayTime then
        self.dayTime = val
        self:applyValues()
    end
end

function DayNightController:setAmbient(val)
    if val ~= self.ambient then
        self.ambient = val
        self:applyValues()
    end
end

function DayNightController:applyValues()
    local scene = self:entity():parentScene()

    if scene then
        local skyColor = self.ambient and (self.dayTime and self.daySkyColor or self.nightSkyColor) or color.black()
        scene:setBackColor(skyColor)
        scene:setAmbientLight(skyColor)
    end

    if self.light then
        self.light:setColor(self.dayTime and self.dayLightColor or self.nightLightColor)
        self.light:setIntensity(self.dayTime and self.dayLightIntensity or self.nightLightIntensity)
    end
end